src/devices/bus/segaai/soundbox.cpp

```diff
@@ -0,0 +1,320 @@
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW4")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW5")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW6")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW7")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_UNUSED)
+INPUT_PORTS_END
+
+ioport_constructor segaai_soundbox_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(soundbox);
+}
+
+void segaai_soundbox_device::device_start()
+{
+	m_row = 0;
+	m_ram.resize(0x20000);
+
+	save_item(NAME(m_ram));
+	save_item(NAME(m_row));
+
+	mem_space().install_ram(0x20000, 0x3ffff, &m_ram[0]);
+	mem_space().install_rom(0x80000, 0x8ffff, m_rom);
+	io_space().install_read_handler(0x20, 0x23, emu::rw_delegate(*m_ym2151, FUNC(ym2151_device::read)));
+	io_space().install_write_handler(0x20, 0x23, emu::rw_delegate(*m_ym2151, FUNC(ym2151_device::write)));
+	io_space().install_read_handler(0x24, 0x27, emu::rw_delegate(*m_tmp8253, FUNC(pit8253_device::read)));
+	io_space().install_write_handler(0x24, 0x27, emu::rw_delegate(*m_tmp8253, FUNC(pit8253_device::write)));
+	io_space().install_read_handler(0x28, 0x2b, emu::rw_delegate(*m_tmp8255, FUNC(i8255_device::read)));
+	io_space().install_write_handler(0x28, 0x2b, emu::rw_delegate(*m_tmp8255, FUNC(i8255_device::write)));
+}
+
+u8 segaai_soundbox_device::tmp8255_porta_r()
+{
+	// Read pressed keys on music keyboard row (see routine @0x82399)
+	u8 result = 0xff;
+	for (int i = 0; i < 8; i++)
+		if (BIT(m_row, i)) result &= m_rows[i]->read();
+	return result;
+}
+
+u8 segaai_soundbox_device::tmp8255_portb_r()
+{
+	return 0xff;
+}
```

Comment:
Why not just use `.set_constant(0xff)` in machine configuration rather than going through this function?

Reply:
Changed to use set_constant.